File: Zds/Zds.Flow/Machinery/Objects/SyncSink.vb
Imports Zds.Flow.Updatables

Namespace Machinery.Objects
    Public Class SyncSink(Of Input)
        Inherits SyncObject
        Implements ISink(Of Input)
        Private _Sink As Core.SyncSink(Of Input)
        Public Property SinkDelegate As Core.Sink(Of Input).SinkDelegate
        Public Property Recursive As Boolean
            Get
                Return _Sink.Recursive
            End Get
            Set(value As Boolean)
                _Sink.Recursive = value
            End Set
        End Property
        Public Property Queue As Collections.IQueue(Of Input)
            Get
                Return _Sink.Queue
            End Get
            Set(value As Collections.IQueue(Of Input))
                _Sink.Queue = value
            End Set
        End Property
        Public Function Receive(obj As Input) As Boolean Implements ISink(Of Input).Receive
            Return _Sink.Receive(obj)
        End Function
        Protected Overrides Sub SyncUpdate()
            MyBase.SyncUpdate()
            _Sink.Activate()
        End Sub
        Private Function InternalSink(Input As Input) As Boolean
            Try
                Return Sink(Input)
            Catch ex As Exception
                Handle(ex)
                Return False
            End Try
        End Function
        Protected Overridable Function Sink(Input As Input) As Boolean
            If SinkDelegate IsNot Nothing Then Return SinkDelegate(Input) Else Return False
        End Function
        Sub New()
            _Sink = New Core.SyncSink(Of Input)
            _Sink.Sink = AddressOf InternalSink
        End Sub
        Sub New(Sink As Core.Sink(Of Input).SinkDelegate)
            Me.New()
            SinkDelegate = Sink
        End Sub
        Sub New(Updater As Updaters.IUpdater)
            MyBase.New(Updater)
            _Sink = New Core.SyncSink(Of Input)
            _Sink.Sink = AddressOf InternalSink
        End Sub
        Sub New(Updater As Updaters.IUpdater, Sink As Core.Sink(Of Input).SinkDelegate)
            Me.New(Updater)
            SinkDelegate = Sink
        End Sub
    End Class
End Namespace

File: Flow/Flow/Machinery/SyncSink.cs
﻿using Flow.Collections;
using Flow.Interfaces;
using Flow.Updatables;
using Flow.Updaters;
using System;
using System.Collections.Generic;
using System.Text;

namespace Flow.Machinery
{
    public class SyncSink<T> : SyncObject, ISyncSink<T>
    {
        private readonly IQueue<T> queue;
        public event SinkDelegate<T> OnSink;

        public override void Reset()
        {
            base.Reset();
            Sink<T>.Clear(queue);
        }

        protected override void SyncUpdate()
        {
            base.SyncUpdate();
            Sink();
        }
        protected virtual void Sink() => OnSink?.Invoke(this);

        public bool Send(T obj) => Sink<T>.Send(queue, obj);
        public int Send(T[] array, int index, int length) => Sink<T>.Send(queue, array, index, length);
        public int Send(T[] array) => Sink<T>.Send(queue, array);

        protected bool Receive(ref T obj) => Sink<T>.Receive(queue, ref obj);
        protected int Receive(T[] array, int index, int length) => Sink<T>.Receive(queue, array, index, length);
        protected int Receive(T[] array) => Sink<T>.Receive(queue, array);

        bool IUpdatableSink<T>.Receive(ref T obj) => Receive(ref obj);
        int IUpdatableSink<T>.Receive(T[] array) => Receive(array);
        int IUpdatableSink<T>.Receive(T[] array, int index, int length) => Receive(array, index, length);

        public SyncSink() => queue = new DynamicArray<T>(1024 * 1024);
        public SyncSink(IQueue<T> queue) => this.queue = queue;
        public SyncSink(IUpdater updater) : base(updater) => queue = new DynamicArray<T>(1024 * 1024);
        public SyncSink(IUpdater updater, IQueue<T> queue) : base(updater) => this.queue = queue;
    }

    public interface ISyncSink<T> : ISyncObject, IUpdatableSink<T> { }
}
